tanglescope/tanglewidthcollector.cpp

```diff
@@ -0,0 +1,88 @@
+#include "iota/tanglescope/tanglewidthcollector.hpp"
+#include <glog/logging.h>
+#include <iota/tanglescope/common/tangledb.hpp>
+#include <unordered_map>
+
+namespace iota {
+namespace tanglescope {
+
+std::map<std::string, std::string> TangleWidthCollector::nameToDescGauges = {
+    {TangleWidthCollector::TANGLE_WIDTH,
+     "Number of transactions approving a single transaction as observed "
+     "across multiple nodes"}};
+
+bool TangleWidthCollector::parseConfiguration(const YAML::Node& conf) {
+  if (!PrometheusCollector::parseConfiguration(conf)) {
+    return false;
+  }
+
+  if (!conf[SNAPSHOT_INTERVAL] || !conf[MEASURE_LINE_AGE]) {
+    return false;
+  }
+
+  _snapshotInterval = conf[SNAPSHOT_INTERVAL].as<uint32_t>();
+  _mesaureLineAge = conf[MEASURE_LINE_AGE].as<uint32_t>();
+
+  return true;
+}
+
+void TangleWidthCollector::collect() {
+  using namespace prometheus;
+  VLOG(3) << __FUNCTION__;
+
+  Exposer exposer{_prometheusExpURI};
+  auto registry = std::make_shared<Registry>();
+  exposer.RegisterCollectable(registry);
+
+  _gauges =
+      buildGaugeMap(registry, "tanglewidthcollector", {}, nameToDescGauges);
+
+  analyzeWidthPeriodically();
+}
+
+void TangleWidthCollector::analyzeWidthPeriodically() {
+  auto pubThread = rxcpp::schedulers::make_current_thread();
+  auto pubWorker = pubThread.create_worker();
+
+  auto& thisRef = *this;
+  if (_snapshotInterval > 0) {
+    pubWorker.schedule_periodically(
+        pubThread.now() + std::chrono::seconds(_mesaureLineAge + 120),
+        std::chrono::seconds(_snapshotInterval),
+        [&thisRef](auto scbl) { thisRef.analyzeWidthImpl(); });
```

Comment:
This is a possible race condition: the class can be destructed before the thread is.
You should keep a reference to this thread & worker as a class member so you can destruct safely

Reply:
This might also apply to other collectors?

Reply:
it might, though in practice tanglescope runner make sure not to exit the program before all collectors return, but we could implement a destructor to force exit